File: packages/kotlin/src/main/kotlin/com/ecollect/sdk/modules/CustomersModule.kt
package com.ecollect.sdk.modules

import com.ecollect.sdk.Config
import com.ecollect.sdk.exceptions.*
import com.ecollect.sdk.types.*
import com.ecollect.sdk.utils.HttpClient

class CustomersModule(
    private val config: Config,
    private val httpClient: HttpClient,
    private val sessionModule: SessionModule
) {
    /**
     * Get or create a CustomerId for a customer.
     * If the customer already exists, the same CustomerId is returned.
     * CustomerId can be used to reduce payload size on subsequent tokenizations.
     *
     * @return The CustomerId string
     */
    suspend fun getOrCreateCustomerId(
        usermail: String,
        cardHolderId: String,
        cardHolderName: String,
        cardHolderIdType: String,
        mobileCountryCode: String,
        mobileNumber: String
    ): String {
        val token = sessionModule.getSessionToken()
        val customerInfoArray = listOf(
            PaymentInfoType(6, "Usermail", usermail),
            PaymentInfoType(17, "CardHolderName", cardHolderName),
            PaymentInfoType(18, "CardHolderIdType", cardHolderIdType),
            PaymentInfoType(19, "CardHolderId", cardHolderId),
            PaymentInfoType(7, "MobileCountryCode", mobileCountryCode),
            PaymentInfoType(8, "MobileNumber", mobileNumber)
        )

        val request = GetCustomerIdRequest(
            entityCode = config.etyCode,
            sessionToken = token,
            customerInfoArray = customerInfoArray
        )
        val url = "${config.baseUrl}getCustomerId"
        val response: GetCustomerIdResponse = httpClient.post(url, request)
        return handleResponse(response, token, request)
    }

    /**
     * Update a customer's information.
     */
    suspend fun updateCustomer(
        customerId: String,
        usermail: String? = null,
        cardHolderName: String? = null,
        cardHolderIdType: String? = null,
        mobileCountryCode: String? = null,
        mobileNumber: String? = null
    ): String {
        val token = sessionModule.getSessionToken()
        val customerInfoArray = mutableListOf(
            PaymentInfoType(attributeCode = 37, attributeDesc = "CustomerId", attributeValue = customerId)
        )
        usermail?.let { customerInfoArray.add(PaymentInfoType(6, "Usermail", it)) }
        cardHolderName?.let { customerInfoArray.add(PaymentInfoType(17, "CardHolderName", it)) }
        cardHolderIdType?.let { customerInfoArray.add(PaymentInfoType(18, "CardHolderIdType", it)) }
        mobileCountryCode?.let { customerInfoArray.add(PaymentInfoType(7, "MobileCountryCode", it)) }
        mobileNumber?.let { customerInfoArray.add(PaymentInfoType(8, "MobileNumber", it)) }

        val request = GetCustomerIdRequest(
            entityCode = config.etyCode,
            sessionToken = token,
            customerInfoArray = customerInfoArray
        )
        val url = "${config.baseUrl}getCustomerId"
        val response: GetCustomerIdResponse = httpClient.post(url, request)
        return handleResponse(response, token, request)
    }

    private suspend fun handleResponse(
        response: GetCustomerIdResponse,
        token: String,
        originalRequest: GetCustomerIdRequest
    ): String {
        return when (response.returnCode) {
            "SUCCESS" -> {
                response.customerInfoArray
                    ?.firstOrNull { it.attributeDesc == "CustomerId" }
                    ?.attributeValue
                    ?: throw CustomerNotFoundException("CustomerId not found in response")
            }
            "FAIL_APIEXPIREDSESSION" -> {
                sessionModule.invalidate()
                val newToken = sessionModule.getSessionToken()
                val retryResponse: GetCustomerIdResponse = httpClient.post(
                    "${config.baseUrl}getCustomerId",
                    originalRequest.copy(sessionToken = newToken)
                )
                handleResponse(retryResponse, newToken, originalRequest)
            }
            "FAIL_INVALIDENTITYCODE" -> throw InvalidConfigException("EntityCode is invalid")
            "FAIL_ACCESSDENIED" -> throw AuthenticationException("Session token authentication failed")
            "FAIL_CARDHOLDERID" -> throw ValidationException("CardHolderId is invalid")
            "FAIL_CARDHOLDERNAME" -> throw ValidationException("CardHolderName is invalid")
            "FAIL_CARDHOLDERIDTYPE" -> throw ValidationException("CardHolderIdType is not valid for this country")
            "FAIL_MOBILECOUNTRYCODE" -> throw ValidationException("MobileCountryCode is invalid")
            "FAIL_MOBILENUMBER" -> throw ValidationException("MobileNumber is invalid")
            "FAIL_MAILFORMAT" -> throw ValidationException("Usermail format is invalid")
            "FAIL_SYSTEM" -> throw NetworkRetryableException("ecollect system error during getCustomerId")
            else -> throw EcollectException("Unexpected returnCode: ${response.returnCode}")
        }
    }
}

File: packages/php/src/Modules/CustomersModule.php
<?php

declare(strict_types=1);

namespace Ecollect\Modules;

use Ecollect\Config;
use Ecollect\Exceptions\CustomerException;
use Ecollect\Utils\ErrorMapper;
use Ecollect\Utils\HttpClient;

/**
 * Manages customer IDs for tokenisation (getOrCreate and update).
 */
class CustomersModule
{
    /** @var Config */
    private $config;

    /** @var HttpClient */
    private $http;

    /** @var SessionModule */
    private $session;

    public function __construct(Config $config, HttpClient $http, SessionModule $session)
    {
        $this->config  = $config;
        $this->http    = $http;
        $this->session = $session;
    }

    /**
     * Get or create a CustomerId for a customer.
     *
     * @param  array<string, string> $customerInfo Keys: email, document_number, document_type, full_name, mobile_country_code, mobile_number
     * @return array<string, mixed>
     * @throws CustomerException
     */
    public function getOrCreateCustomerId(array $customerInfo): array
    {
        $url          = $this->config->baseUrl() . '/getCustomerId';
        $sessionToken = $this->session->getActive();

        $customerInfoArray = [];

        $add = function (int $code, string $desc, ?string $value) use (&$customerInfoArray): void {
            if ($value !== null && $value !== '') {
                $customerInfoArray[] = ['AttributeCode' => $code, 'AttributeDesc' => $desc, 'AttributeValue' => $value];
            }
        };

        $add(6,  'Usermail',          $customerInfo['email']               ?? null);
        $add(19, 'CardHolderId',      $customerInfo['document_number']     ?? null);
        $add(18, 'CardHolderIdType',  $customerInfo['document_type']       ?? null);
        $add(17, 'CardHolderName',    $customerInfo['full_name']           ?? null);
        $add(7,  'MobileCountryCode', $customerInfo['mobile_country_code'] ?? null);
        $add(8,  'MobileNumber',      $customerInfo['mobile_number']       ?? null);

        $body = [
            'EntityCode'         => $this->config->etyCode,
            'SessionToken'       => $sessionToken,
            'CustomerInfoArray'  => $customerInfoArray,
        ];

        $res = $this->http->post($url, $body);

        if (($res['ReturnCode'] ?? '') === 'FAIL_APIEXPIREDSESSION') {
            $this->session->invalidate();
            $body['SessionToken'] = $this->session->getActive();
            $res                  = $this->http->post($url, $body);

            if (($res['ReturnCode'] ?? '') !== 'SUCCESS') {
                throw ErrorMapper::map($res['ReturnCode'] ?? 'FAIL_SYSTEM');
            }
        } elseif (($res['ReturnCode'] ?? '') !== 'SUCCESS') {
            throw ErrorMapper::map($res['ReturnCode'] ?? 'FAIL_SYSTEM');
        }

        return $this->parseCustomer($res, $customerInfo);
    }

    /**
     * Update customer information by providing a CustomerId.
     *
     * @param  string                $customerId
     * @param  array<string, string> $updatedInfo
     * @return array<string, mixed>
     */
    public function updateCustomerInfo(string $customerId, array $updatedInfo): array
    {
        $url          = $this->config->baseUrl() . '/getCustomerId';
        $sessionToken = $this->session->getActive();

        $items = [
            ['AttributeCode' => 100, 'AttributeDesc' => 'CustomerId', 'AttributeValue' => $customerId],
        ];

        $add = function (int $code, string $desc, ?string $value) use (&$items): void {
            if ($value !== null && $value !== '') {
                $items[] = ['AttributeCode' => $code, 'AttributeDesc' => $desc, 'AttributeValue' => $value];
            }
        };

        $add(6,  'Usermail',          $updatedInfo['email']               ?? null);
        $add(17, 'CardHolderName',    $updatedInfo['full_name']           ?? null);
        $add(18, 'CardHolderIdType',  $updatedInfo['document_type']       ?? null);
        $add(7,  'MobileCountryCode', $updatedInfo['mobile_country_code'] ?? null);
        $add(8,  'MobileNumber',      $updatedInfo['mobile_number']       ?? null);

        $body = [
            'EntityCode'        => $this->config->etyCode,
            'SessionToken'      => $sessionToken,
            'CustomerInfoArray' => $items,
        ];

        $res = $this->http->post($url, $body);

        if (($res['ReturnCode'] ?? '') !== 'SUCCESS') {
            throw ErrorMapper::map($res['ReturnCode'] ?? 'FAIL_SYSTEM');
        }

        return $this->parseCustomer($res, $updatedInfo);
    }

    /**
     * @param  array<string, mixed>  $res
     * @param  array<string, string> $original
     * @return array<string, mixed>
     * @throws CustomerException
     */
    private function parseCustomer(array $res, array $original): array
    {
        $arr = $res['CustomerInfoArray'] ?? [];

        $findAttr = function (int $code) use ($arr): ?string {
            foreach ($arr as $item) {
                if ((int)($item['AttributeCode'] ?? -1) === $code) {
                    return $item['AttributeValue'] ?? null;
                }
            }
            return null;
        };

        $customerId = $findAttr(100);
        if (!$customerId) {
            throw new CustomerException('getCustomerId response did not return a CustomerId');
        }

        return [
            'customerId'         => $customerId,
            'email'              => $original['email']               ?? null,
            'fullName'           => $original['full_name']           ?? null,
            'documentType'       => $original['document_type']       ?? null,
            'documentNumber'     => $original['document_number']     ?? null,
            'mobileCountryCode'  => $original['mobile_country_code'] ?? null,
            'mobileNumber'       => $original['mobile_number']       ?? null,
        ];
    }
}
